src/dune_engine/build_system.ml

```diff
@@ -1551,10 +1528,16 @@ end = struct
           Action.progn [ action; Action.write_file stamp_file "" ]
     in
     let* () =
-      Fiber.parallel_iter_set
-        (module Path.Set)
-        chdirs
-        ~f:(fun p -> Memo.Build.run (Fs.mkdir_p_or_assert_existence ~loc p))
+      let chdirs = Action.chdirs action in
```

Comment:
What about pushing the check to `Action.chdirs`? Or even to when the action itself is constructed. I think that `Action.chdirs` should return `Path.Build.Set.t` as well.

Reply:
As I wrote in the description, that seemed more invasive to me: right now, we are saying that actions in the rules that we execute can't `chdir` to non-build directories. If we push the check into `Action.chdirs`, we're saying that there can be no such actions altogether. Do we really want that? What if there is some action we create on the side to somehow execute outside of `build_system.ml`?